features/custom_codecs/README.md

```diff
@@ -374,6 +374,212 @@ Foo<Bar> foo = row.get(0, new TypeToken<Foo<Bar>>(){})
 Foo<Bar> foo = row.get(0, Foo.class);
 ```
 
+### Support for subtype polymorphism
+
+Suppose the following class hierarchy:
+
+```java
+class A {}
+class B extends A {}
+```
+
+By default, a codec will accept to serialize any object that extends or
+implements its declared Java type: a codec such as
+`ACodec extends TypeCodec<A>` will accept `B` instances as well.
+
+This allows a codec to handle interfaces and superclasses
+in a generic way, regardless of the actual implementation being
+used by client code; for example, the driver has a built-in codec
+that handles `List` instances, and this codec is capable of
+serializing any concrete `List` implementation.
+
+But this has one caveat: when setting or retrieving values
+with `get()` and `set()`, *it is vital to pass the exact 
+Java type the codec handles*.
+
+The following example works:
+
+```java
+codecRegistry.register(new ACodec());
+BoundStatement bs = ...;
+bs.set(0, new A(), A.class);
+bs.set(0, new B(), A.class); // ACodec accepts B, but is declared with A
+```
+
+But the following doesn't:
+
+```java
+codecRegistry.register(new ACodec());
+BoundStatement bs = ...;
+bs.set(0, new B(), B.class); // throws CodecNotFoundException
+```
+
+The same is valid when retrieving values:
+
+```java
+codecRegistry.register(new ACodec());
+Row row = ...
+A a = row.get(0, A.class); // works
+B b = row.get(0, B.class); // throws CodecNotFoundException
+```
+
+Another anti-pattern that is strongly discouraged is to register a codec for
+a Java type and another codec for one of its superclasses or superinterfaces.
+In this situation, the registry will choose the first registered one,
+*even if the other one is a better match*.
+
+For example, if there are registered codecs for both `A` and `B`
+(i.e. `ACodec extends TypeCodec<A>` and `BCodec extends TypeCodec<B>`),
+when looking for a codec accepting `A` instances, which codec the registry will choose
+will depend on their registration order. This can lead to unpredictable results
+and runtime errors, and thus should always be avoided.
```

Comment:
Well that's only true if you search for a codec for a value, right? Now that the type-based methods are invariant, B.class will return the BCodec and A.class the ACodec, whatever the order.